### PR TITLE
SonarLint should not show issues marked as FP, Won't Fix or Fixed

### PR DESCRIPTION
**Pull request by Evangelink**

### CODE REVIEW

File: src/Integration.Vsix/Suppression/LiveIssueFactory.cs [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+            string projectId;
+            if (!projectPathToProjectIdMap.TryGetValue(project.FilePath, out projectId))
+            {
+                Debug.Fail($"Expecting to have a project id for the Roslyn project: {project.FilePath}");
+                return null;
+            }
+
+            // Get the whole of the line of text containing the issue (needed to compute the checksum)
+            // TODO: checksum for file level issues?
```
Comment (SonarTech) [inline src/Integration.Vsix/Suppression/LiveIssueFactory.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration.Vsix/Suppression/LiveIssueFactory.cs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        }
+
+        private void BuildProjectPathToIdMap()
+        {
+            projectPathToProjectIdMap = new Dictionary<string, string>(StringComparer.InvariantCultureIgnoreCase);
+
+            IVsSolution solution = this.serviceProvider.GetService<SVsSolution, IVsSolution>();
+            Debug.Assert(solution != null, "Cannot find SVsSolution");
+
+            // TODO: handle return codes
```
Comment (SonarTech) [inline src/Integration.Vsix/Suppression/LiveIssueFactory.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration.Vsix/Suppression/LiveIssueFactory.cs (hunk 1 above)
Comment (SonarTech) [inline src/Integration.Vsix/Suppression/LiveIssueFactory.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration.Vsix/Helpers/FileUtilities.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            [In] FileAttributes dwAttrTo);
+
+        /// <summary>
+        /// Returns the relative path to the second file from the first, or 
+        /// <paramref name="toFilePath"/> if the paths are not relative.
+        /// </summary>
+        public static string GetRelativePath(string fromFilePath, string toFilePath)
+        {
+            var sb = new StringBuilder(260);
+            bool success = PathRelativePathTo(sb, fromFilePath, FileAttributes.Normal, toFilePath, FileAttributes.Normal);
```
Comment (SonarTech) [inline src/Integration.Vsix/Helpers/FileUtilities.cs]: MAJOR Verify that this is the intended value; it is the same as the 3rd argument. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS4142)

File: src/Integration.Vsix/Suppression/LiveIssueFactory.cs (hunk 2 above)
Comment (SonarTech) [inline src/Integration.Vsix/Suppression/LiveIssueFactory.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration.Vsix/Suppression/SuppressionManager.cs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            // TODO: ?need to make file path relative to the project file path
+            // As a minimum, the project, file and rule id must match
+            var issuesInFile = sonarqubeIssueProvider.GetSuppressedIssues(liveIssue.ProjectGuid, liveIssue.IssueFilePath);
+
+            if (issuesInFile == null)
+            {
+                return true;
+            }
+
+            // TODO: rule repository?
```
Comment (SonarTech) [inline src/Integration.Vsix/Suppression/SuppressionManager.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration.Vsix/Helpers/FileUtilities.cs (hunk 3 above)
Comment (SonarTech) [inline src/Integration.Vsix/Helpers/FileUtilities.cs]: MAJOR Verify that this is the intended value; it is the same as the 3rd argument. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS4142)

File: src/Integration.Vsix/Suppression/SuppressionManager.cs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+            if (liveIssue == null)
+            {
+                return true; // Unable to get the data required to map a Roslyn issue to a SonarQube issue
+            }
+
+            // Issues match if:
+            // 1. Same component, same file, same error code, same line hash        // tolerant to line number changing
+            // 2. Same component, same file, same error code, same line             // tolarant to code on the line changing e.g. var rename
+
+            // TODO: ?need to make file path relative to the project file path
```
Comment (SonarTech) [inline src/Integration.Vsix/Suppression/SuppressionManager.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration.Vsix/Suppression/SuppressionManager.cs (hunk 4 above)
Comment (SonarTech) [inline src/Integration.Vsix/Suppression/SuppressionManager.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration/Suppression/SonarQubeIssuesProvider.cs [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            // TODO: may not be connected
+            if (!this.sonarQubeService.IsConnected)
+            {
+                return;
+            }
+
+            cancellationTokenSource?.Cancel();
+            cancellationTokenSource = new CancellationTokenSource();
+
+            // TODO: Handle race conditions
```
Comment (SonarTech) [inline src/Integration/Suppression/SonarQubeIssuesProvider.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration.Vsix/Suppression/SuppressionManager.cs (hunk 5 above)
Comment (SonarTech) [inline src/Integration.Vsix/Suppression/SuppressionManager.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration/Suppression/SonarQubeIssuesProvider.cs [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        private CancellationTokenSource cancellationTokenSource;
+
+        [ImportingConstructor]
+        internal SonarQubeIssuesProvider(ISonarQubeService sonarQubeService, IActiveSolutionBoundTracker solutionBoundTacker)
+        {
+            this.sonarQubeService = sonarQubeService;
+            this.solutionBoundTacker = solutionBoundTacker;
+            this.solutionBoundTacker.SolutionBindingChanged += OnSolutionBoundChanged;
+
+            // TODO: Use mockable timer
```
Comment (SonarTech) [inline src/Integration/Suppression/SonarQubeIssuesProvider.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration/Suppression/SonarQubeIssuesProvider.cs (hunk 6 above)
Comment (SonarTech) [inline src/Integration/Suppression/SonarQubeIssuesProvider.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration/Suppression/SonarQubeIssuesProvider.cs [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+            }
+            else
+            {
+                refreshTimer.Stop();
+            }
+        }
+
+        private async Task SynchronizeSuppressedIssues()
+        {
+            // TODO: may not be connected
```
Comment (SonarTech) [inline src/Integration/Suppression/SonarQubeIssuesProvider.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration/Suppression/SonarQubeIssuesProvider.cs (hunk 7 above)
Comment (SonarTech) [inline src/Integration/Suppression/SonarQubeIssuesProvider.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration/Suppression/SonarQubeIssuesProvider.cs [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+            }
+
+            refreshTimer.Dispose();
+            this.solutionBoundTacker.SolutionBindingChanged -= OnSolutionBoundChanged;
+            this.isDisposed = true;
+        }
+
+        public IEnumerable<SonarQubeIssue> GetSuppressedIssues(string projectGuid, string filePath)
+        {
+            // TODO: Block the call while the cache is being built + handle multi-threading
```
Comment (SonarTech) [inline src/Integration/Suppression/SonarQubeIssuesProvider.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration/Suppression/SonarQubeIssuesProvider.cs (hunk 8 above)
Comment (SonarTech) [inline src/Integration/Suppression/SonarQubeIssuesProvider.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration/Suppression/SonarQubeIssuesProvider.cs [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+            // TODO: ensure we've got data to enable end to end testing
```
Comment (SonarTech) [inline src/Integration/Suppression/SonarQubeIssuesProvider.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration/Suppression/SonarQubeIssuesProvider.cs (hunk 9 above)
Comment (SonarTech) [inline src/Integration/Suppression/SonarQubeIssuesProvider.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration/Suppression/SonarQubeIssuesProvider.cs (hunk 10 above)
Comment (SonarTech) [inline src/Integration/Suppression/SonarQubeIssuesProvider.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 10 issues
* ![MAJOR][MAJOR] 1 major
* ![INFO][INFO] 9 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: src/Integration.Vsix/Helpers/FileUtilities.cs [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+            [In] string pszTo,
+            [In] FileAttributes dwAttrTo);
+
+        /// <summary>
+        /// Returns the relative path to the second file from the first, or 
+        /// <paramref name="toFilePath"/> if the paths are not relative.
+        /// </summary>
+        public static string GetRelativePath(string fromFilePath, string toFilePath)
+        {
+            var sb = new StringBuilder(260);
```
Comment (michalb-sonar) [inline src/Integration.Vsix/Helpers/FileUtilities.cs]: Put 260 into a constant. Maybe MAX_PATH or MaxPathLength?

File: src/Integration.Vsix/Helpers/FileUtilities.cs (hunk 3 above)
Comment (michalb-sonar) [reply src/Integration.Vsix/Helpers/FileUtilities.cs]: Should we not report here? E.g. when two values are enum consts and not next to each other, WDYT?

File: src/Integration.Vsix/Helpers/FileUtilities.cs [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+            bool success = PathRelativePathTo(sb, fromFilePath, FileAttributes.Normal, toFilePath, FileAttributes.Normal);
+
+            if (success)
+            {
+                // Strip of the the leading ".\" (i.e. if the files are in the same directory)
+                int offset = 0;
+                if (sb[0] == '.' && sb[1] == '\\')
```
Comment (michalb-sonar) [inline src/Integration.Vsix/Helpers/FileUtilities.cs]: I think it's ok, but just to double-check: is the return string always longer than 2 characters?

File: src/Integration.Vsix/Suppression/LiveIssueFactory.cs [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+            // 1. Call with nulls to get the number of files
```
Comment (michalb-sonar) [inline src/Integration.Vsix/Suppression/LiveIssueFactory.cs]: That API is so C!

File: src/Integration.Vsix/Suppression/LiveIssueFactory.cs [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+            uint fileCount;
+            solution.GetProjectFilesInSolution((uint)__VSGETPROJFILESFLAGS.GPFF_SKIPUNLOADEDPROJECTS, 0, null, out fileCount);
+
+            // 2. Size array and call again to get the data
+            string[] fileNames = new string[fileCount];
+            solution.GetProjectFilesInSolution((uint)__VSGETPROJFILESFLAGS.GPFF_SKIPUNLOADEDPROJECTS, fileCount, fileNames, out fileCount);
+
+            IVsSolution5 soln5 = (IVsSolution5)solution;
```
Comment (michalb-sonar) [inline src/Integration.Vsix/Suppression/LiveIssueFactory.cs]: var?

File: src/Integration.Vsix/Suppression/SuppressionManager.cs [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+
+        private void OnSolutionBindingChanged(object sender, ActiveSolutionBindingEventArgs e)
+        {
+            RefreshSuppresionHandling();
+        }
+
+        private bool ShouldIssueBeReported(Diagnostic diagnostic)
+        {
+            // This method is called for every analyzer issue that is raised so it should be fast.
+            if (!diagnostic.Location.IsInSource) { return true; }
```
Comment (michalb-sonar) [inline src/Integration.Vsix/Suppression/SuppressionManager.cs]: You could merge these two ifs and move brackets into a separate line.

File: src/Integration/Suppression/SonarQubeIssuesProvider.cs [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+            if (solutionBoundTacker.IsActiveSolutionBound &&
+                this.cachedSuppressedIssues == null)
+            {
+                SynchronizeSuppressedIssues().Wait(30000);
```
Comment (michalb-sonar) [inline src/Integration/Suppression/SonarQubeIssuesProvider.cs]: Make time a const, just like MillisecondsToWaitBetweenRefresh

**Pull request closed without merging**